dev/src/index.ts

```diff
@@ -850,87 +856,51 @@ export class Firestore {
   behavior.`);
       }
-      this._clientInitialized = this._initClientPool().then(clientPool => {
-        this._clientPool = clientPool;
-      });
     }
 
-    return this._clientInitialized!.then(() => this._clientPool!.run(op));
-  }
-
-  /**
-   * Initializes the client pool and invokes Project ID detection. Returns a
-   * Promise on completion.
-   *
-   * @private
-   */
-  private _initClientPool(): Promise<ClientPool<GapicClient>> {
-    assert(!this._clientInitialized, 'Client pool already initialized');
-
-    const clientPool =
-        new ClientPool(MAX_CONCURRENT_REQUESTS_PER_CLIENT, () => {
-          const client = new module.exports.v1(this._settings);
-          logger('Firestore', null, 'Initialized Firestore GAPIC Client');
-          return client;
-        });
-
-    const projectIdProvided =
-        this._referencePath!.projectId !== '{{projectId}}';
+    this._settingsFrozen = true;
 
-    if (projectIdProvided) {
-      return Promise.resolve(clientPool);
-    } else {
-      return clientPool.run(client => this._detectProjectId(client))
-          .then(projectId => {
-            this._referencePath =
-                new ResourcePath(projectId, this._referencePath!.databaseId);
-            return clientPool;
+    if (this._projectId === undefined) {
+      this._projectId = await this._clientPool.run(gapicClient => {
+        return new Promise((resolve, reject) => {
+          gapicClient.getProjectId((err: Error, projectId: string) => {
+            if (err) {
+              logger(
+                  'Firestore._detectProjectId', null,
+                  'Failed to detect project ID: %s', err);
+              reject(err);
+            } else {
+              logger(
+                  'Firestore._detectProjectId', null, 'Detected project ID: %s',
+                  projectId);
+              resolve(projectId);
+            }
           });
+        });
+      });
     }
   }
 
   /**
-   * Auto-detects the Firestore Project ID.
-   *
+   * Ensures that the client has been fully initialized;
    * @private
-   * @param gapicClient The Firestore GAPIC client.
-   * @return A Promise that resolves with the Project ID.
    */
-  private _detectProjectId(gapicClient: GapicClient): Promise<string> {
-    return new Promise((resolve, reject) => {
-      gapicClient.getProjectId((err: Error, projectId: string) => {
-        if (err) {
-          logger(
-              'Firestore._detectProjectId', null,
-              'Failed to detect project ID: %s', err);
-          reject(err);
-        } else {
-          logger(
-              'Firestore._detectProjectId', null, 'Detected project ID: %s',
-              projectId);
-          resolve(projectId);
-        }
-      });
-    });
+  private ensureClientInitialized(): void {
+    if (this._projectId === undefined) {
+      // Note that we do not mention `initializeIfNeeded()` here to not leak the
+      // private API.
```

Comment:
I removed it. I do however thing that we should not mention `initializeIfNeeded()`. If we point out that we should have called `initializeIfNeeded()`, our users will do so instead of filing a GitHub issue.